File: lib/engine/world_position.dart
import 'dart:ui' show Offset;

import 'package:batufo/dart_types.dart';
import 'package:batufo/engine/tile_position.dart';

@immutable
class WorldPosition {
  final double x;
  final double y;

  const WorldPosition(this.x, this.y);

  static WorldPosition fromTilePosition(TilePosition tp) {
    assert(tileSize != null, 'init tileSize first');
    final t = tileSize;
    final x = t * tp.col + tp.relX;
    final y = t * tp.row + tp.relY;
    return WorldPosition(x, y);
  }

  TilePosition toTilePosition() {
    assert(tileSize != null, 'init tileSize first');
    final t = tileSize;
    final col = x ~/ t;
    final row = y ~/ t;
    final relX = x % t;
    final relY = y % t;
    return TilePosition(col, row, relX, relY);
  }

  Offset toOffset() {
    return Offset(x, y);
  }

  static WorldPosition fromOffset(Offset offset) {
    return WorldPosition(offset.dx, offset.dy);
  }

  static double tileSize;
}

File: backend/packages/core/src/arena/world-position.ts
import { TilePosition } from './tile-position'

export class WorldPosition {
  constructor(readonly x: number, readonly y: number) {}

  static fromTilePosition(tp: TilePosition, tileSize: number): WorldPosition {
    const t = tileSize
    const x = t * tp.col + tp.relX
    const y = t * tp.row + tp.relY
    return new WorldPosition(x, y)
  }

  toTilePosition(tileSize: number): TilePosition {
    const t = tileSize
    const col = this.x / t
    const row = this.y / t
    const relX = this.x % t
    const relY = this.y % t
    return new TilePosition(col, row, relX, relY)
  }
}
